File: java/steganography(pic_in_pic)x/src/SteganographyPic.java
import stringmanip.StringManipulations;

import javax.imageio.ImageIO;
import java.awt.*;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.math.BigInteger;
import java.util.ArrayList;
import java.util.List;

public class SteganographyPic {
    private static final List<String> bufferDelimiterList = new ArrayList<String>() {
        {
            add("1010");
            add("1111");
            add("0101");
        }
    };
    private static final List<String> delimiterList = new ArrayList<String>() {
        {
            add("1111");
            add("1010");
            add("1111");
        }
    };
    private static final String padding = "0000";
    private static final int bufferGrouping = 3;
    private static final String separator = System.lineSeparator();

    public static void main(String[] args) {
        System.out.println();
    }

    public static String generateEncodedImage(BufferedImage image, String picturePath) throws IOException {
        final File picturePathFile = new File(picturePath);

        ImageIO.write(image, "png", picturePathFile);
        return picturePathFile.getAbsolutePath();
    }

    private static List<String> completeGroupsForPixelEmbedding(int input) {
        final List<String> reversedBinListNoPadding = StringManipulations.chunkString(StringManipulations.reverseString(Integer.toBinaryString(input)), 4);
        final ArrayList<String> reversedBinList = new ArrayList<>(0);

        for (String s : reversedBinListNoPadding) {
            reversedBinList.add(StringManipulations.padRightWithZeros(s, 4));
        }

        if (reversedBinList.size() % bufferGrouping != 0) {
            final int listLimit = ((reversedBinList.size() / bufferGrouping) + 1) * bufferGrouping;
            for (int i = reversedBinList.size(); i < listLimit; i++) {
                reversedBinList.add(padding);
            }
        }

        final ArrayList<String> binList = new ArrayList<>(0);
        for (int j = reversedBinList.size() - 1; j >= 0; j--) {
            binList.add(StringManipulations.reverseString(reversedBinList.get(j)));
        }

        return binList;
    }

    public static BufferedImage embedMessage(BufferedImage pictureToEncode, BufferedImage pictureToHide) throws OriginalPicTooSmallException, IOException {
        final BufferedImage originalBufferedImg = pictureToEncode;
        final int originalPicWidth = originalBufferedImg.getWidth();
        final int originalPicHeight = originalBufferedImg.getHeight();
        final int originalPicPixelArea = originalPicHeight * originalPicWidth;

        final BufferedImage picToHideBufferedImage = pictureToHide;
        final int picToHideWidth = picToHideBufferedImage.getWidth();
        final int picToHideHeight = picToHideBufferedImage.getHeight();
        final int picToHidePixelArea = picToHideHeight * picToHideWidth;

        final List<String> picToHideWidthBinList = completeGroupsForPixelEmbedding(picToHideWidth);
        final List<String> picToHideHeightBinList = completeGroupsForPixelEmbedding(picToHideHeight);
        final ArrayList<String> picToHidePixelAreaBinList = new ArrayList<String>() {
            {
                addAll(picToHideWidthBinList);
                addAll(bufferDelimiterList);
                addAll(picToHideHeightBinList);
            }
        };
        final int totalPixelsForBuffer = picToHidePixelAreaBinList.size() / bufferGrouping;

        if (picToHidePixelArea > (originalPicPixelArea - totalPixelsForBuffer)) {
            throw new OriginalPicTooSmallException();
        }

        final BufferedImage newEncodedBufferedImage = new BufferedImage(originalPicWidth, originalPicHeight, originalBufferedImg.getType());

        final File file = new File(".\\debugEn.txt");
        final FileOutputStream outputStream = new FileOutputStream(file);

        outputStream.write(("Image in Binary..." + separator).getBytes());

        int pixelsLeftForBuffer = totalPixelsForBuffer;
        boolean insertDelimiter = false;
        int picToHideYIndex = 0;
        int picToHideXIndex = -2;
        boolean shouldEmbedPixel = true;

        for (int yIndex = 0; yIndex < originalPicHeight; yIndex++) {
            for (int xIndex = 0; xIndex < originalPicWidth; xIndex++) {
                final int originalImgPixel = originalBufferedImg.getRGB(xIndex, yIndex);

                if (shouldEmbedPixel) {
                    final Color originalImgPixelColor = new Color(originalImgPixel);
                    final List<String> originalImgPixelRed = StringManipulations.chunkString(StringManipulations.padLeftWithZeros(Integer.toBinaryString(originalImgPixelColor.getRed()), 8),4);
                    final List<String> originalImgPixelGreen = StringManipulations.chunkString(StringManipulations.padLeftWithZeros(Integer.toBinaryString(originalImgPixelColor.getGreen()), 8),4);
                    final List<String> originalImgPixelBlue = StringManipulations.chunkString(StringManipulations.padLeftWithZeros(Integer.toBinaryString(originalImgPixelColor.getBlue()), 8),4);
                    final List<String> originalImgPixelAlpha = StringManipulations.chunkString(StringManipulations.padLeftWithZeros(Integer.toBinaryString(originalImgPixelColor.getAlpha()), 8),4);
                    final String originalImgRgbBinStr = String.join(" | ", StringManipulations.chunkString(Integer.toBinaryString(originalImgPixel), 8));

                    outputStream.write(("Original Image Pixel (x:" + xIndex + ", y:" + yIndex + "): R - " + String.join(" ", originalImgPixelRed) + " G - " + String.join(" ", originalImgPixelGreen) + " B - " + String.join(" ", originalImgPixelBlue) + " A - " + String.join(" ", originalImgPixelAlpha)).getBytes());
                    outputStream.write(separator.getBytes());
                    outputStream.write(("Original Image Pixel (Full): " + originalImgRgbBinStr).getBytes());

                    outputStream.write(separator.getBytes());

                    final List<String> picToHidePixelRed;
                    final List<String> picToHidePixelGreen;
                    final List<String> picToHidePixelBlue;
                    final List<String> picToHidePixelAlpha;
                    final String picToHideRgbBinStr;

                    if (pixelsLeftForBuffer > 0) {
                        final int startIndex = picToHidePixelAreaBinList.size() - (pixelsLeftForBuffer * bufferGrouping);

                        picToHidePixelRed = new ArrayList<String>() {{ add(picToHidePixelAreaBinList.get(startIndex)); add(padding); }};
                        picToHidePixelGreen = new ArrayList<String>() {{ add(picToHidePixelAreaBinList.get(startIndex + 1)); add(padding); }};
                        picToHidePixelBlue = new ArrayList<String>() {{ add(picToHidePixelAreaBinList.get(startIndex + 2)); add(padding); }};
                        picToHidePixelAlpha = originalImgPixelAlpha;
                        picToHideRgbBinStr = String.join(" ", picToHidePixelAlpha) + " | " + String.join(" ", picToHidePixelRed) + " | " + String.join(" ", picToHidePixelGreen) + " | " + String.join(" ", picToHidePixelBlue);

                        outputStream.write(("BUFFER " + separator).getBytes());

                        pixelsLeftForBuffer--;

                        if (pixelsLeftForBuffer == 0) {
                            insertDelimiter = true;
                        }
                    }
                    else if (insertDelimiter) {
                        picToHidePixelRed = new ArrayList<String>() {{ add(delimiterList.get(0)); add(delimiterList.get(0)); }};
                        picToHidePixelGreen = new ArrayList<String>() {{ add(delimiterList.get(1)); add(delimiterList.get(1)); }};
                        picToHidePixelBlue = new ArrayList<String>() {{ add(delimiterList.get(2)); add(delimiterList.get(2)); }};
                        picToHidePixelAlpha = originalImgPixelAlpha;
                        picToHideRgbBinStr = String.join(" ", picToHidePixelAlpha) + " | " + String.join(" ", picToHidePixelRed) + " | " + String.join(" ", picToHidePixelGreen) + " | " + String.join(" ", picToHidePixelBlue);

                        outputStream.write(("DELIITER " + separator).getBytes());
                        insertDelimiter = false;
                    }
                    else {
                        final int picToHidePixel = picToHideBufferedImage.getRGB(picToHideXIndex, picToHideYIndex);
                        final Color picToHidePixelColor = new Color(picToHidePixel);

                        picToHidePixelRed = StringManipulations.chunkString(StringManipulations.padLeftWithZeros(Integer.toBinaryString(picToHidePixelColor.getRed()), 8), 4);
                        picToHidePixelGreen = StringManipulations.chunkString(StringManipulations.padLeftWithZeros(Integer.toBinaryString(picToHidePixelColor.getGreen()), 8), 4);;
                        picToHidePixelBlue = StringManipulations.chunkString(StringManipulations.padLeftWithZeros(Integer.toBinaryString(picToHidePixelColor.getBlue()), 8), 4);;
                        picToHidePixelAlpha = StringManipulations.chunkString(StringManipulations.padLeftWithZeros(Integer.toBinaryString(picToHidePixelColor.getAlpha()), 8), 4);;
                        picToHideRgbBinStr = String.join(" | ", StringManipulations.chunkString(Integer.toBinaryString(picToHidePixel), 8));
                    }

                    outputStream.write(("Picture To Hide Image Pixel (x:" + picToHideXIndex + ", y:" + picToHideYIndex + "): R - " + String.join(" ", picToHidePixelRed) + " G - " + String.join(" ", picToHidePixelGreen) + " B - " + String.join(" ", picToHidePixelBlue) + " A - " + String.join(" ", picToHidePixelAlpha)).getBytes());
                    outputStream.write(separator.getBytes());
                    outputStream.write(("Picture To Hide Image Pixel (Full): " + picToHideRgbBinStr).getBytes());

                    outputStream.write(separator.getBytes());

                    final String newImgPixelRed = String.join("", new ArrayList<String>() {{ add(originalImgPixelRed.get(0)); add(picToHidePixelRed.get(0)); }});
                    final String newImgPixelGreen = String.join("", new ArrayList<String>() {{ add(originalImgPixelGreen.get(0)); add(picToHidePixelGreen.get(0)); }});
                    final String newImgPixelBlue = String.join("", new ArrayList<String>() {{ add(originalImgPixelBlue.get(0)); add(picToHidePixelBlue.get(0)); }});
                    final String newImgPixelAlpha = String.join("", new ArrayList<String>() {{ add(originalImgPixelAlpha.get(0)); add(picToHidePixelAlpha.get(0)); }});
                    final String newImgRgbBinStr = newImgPixelAlpha + newImgPixelRed + newImgPixelGreen + newImgPixelBlue;

                    outputStream.write(("New Image Pixel (x:" + xIndex + ", y:" + yIndex + "): R - " + String.join(" ", StringManipulations.chunkString(newImgPixelRed, 4)) + " G - " + String.join(" ", StringManipulations.chunkString(newImgPixelGreen, 4)) + " B - " + String.join(" ", StringManipulations.chunkString(newImgPixelBlue, 4)) + " A - " + String.join(" ", StringManipulations.chunkString(newImgPixelAlpha, 4))).getBytes());
                    outputStream.write(separator.getBytes());
                    outputStream.write(("New Image Pixel (Full): " + String.join(" | ", StringManipulations.chunkString(newImgRgbBinStr, 8))).getBytes());

                    outputStream.write(separator.getBytes());
                    outputStream.write(separator.getBytes());

                    newEncodedBufferedImage.setRGB(xIndex, yIndex, (new BigInteger(newImgRgbBinStr, 2)).intValue());
                }
                else {
                    newEncodedBufferedImage.setRGB(xIndex, yIndex, originalImgPixel);
                }

                if (picToHideYIndex + 1 == picToHideHeight && picToHideXIndex + 1 == picToHideWidth) {
                    insertDelimiter = true;
                    picToHideYIndex = -100;
                }
                else if (picToHideYIndex < 0)
                    shouldEmbedPixel = false;

                if (picToHideXIndex + 1 == picToHideWidth) {
                    picToHideXIndex = 0;
                    picToHideYIndex++;
                }
                else if (pixelsLeftForBuffer == 0)
                    picToHideXIndex++;
            }
        }

        outputStream.flush();
        outputStream.close();

        return newEncodedBufferedImage;
    }

    public static BufferedImage retrieveEncodedImageFromImage(BufferedImage encodedPic) throws IOException {
        int retrievedImgHeight = 100;
        int retrievedImgWidth = 100;

        final BufferedImage encodedBufferedImg = encodedPic;
        BufferedImage retrievedBufferedImg = new BufferedImage(retrievedImgWidth, retrievedImgHeight, encodedBufferedImg.getType());

        boolean bufferRetrieved = false;
        StringBuilder retrievedBuffer = new StringBuilder();
        int retrievedImgYIndex = 0;
        int retrievedImgXIndex = 0;

        final File file = new File(".\\debugRet.txt");
        final FileOutputStream outputStream = new FileOutputStream(file);

        outer: for (int yIndex = 0; yIndex < encodedBufferedImg.getHeight(); yIndex++) {
            for (int xIndex = 0; xIndex < encodedBufferedImg.getWidth(); xIndex++) {

                final int encodedImgPixel = encodedBufferedImg.getRGB(xIndex, yIndex);
                final Color encodedImgPixelColor = new Color(encodedImgPixel);
                final List<String> encodedImgPixelRed = StringManipulations.chunkString(StringManipulations.padLeftWithZeros(Integer.toBinaryString(encodedImgPixelColor.getRed()), 8),4);
                final List<String> encodedImgPixelGreen = StringManipulations.chunkString(StringManipulations.padLeftWithZeros(Integer.toBinaryString(encodedImgPixelColor.getGreen()), 8),4);
                final List<String> encodedImgPixelBlue = StringManipulations.chunkString(StringManipulations.padLeftWithZeros(Integer.toBinaryString(encodedImgPixelColor.getBlue()), 8),4);
                final List<String> encodedImgPixelAlpha = StringManipulations.chunkString(StringManipulations.padLeftWithZeros(Integer.toBinaryString(encodedImgPixelColor.getAlpha()), 8),4);
                final String encodedImgRgbBinStr = String.join(" | ", StringManipulations.chunkString(Integer.toBinaryString(encodedImgPixel), 8));

                outputStream.write(("Encoded Image Pixel (x:" + xIndex + ", y:" + yIndex + "): R - " + String.join(" ", encodedImgPixelRed) + " G - " + String.join(" ", encodedImgPixelGreen) + " B - " + String.join(" ", encodedImgPixelBlue) + " A - " + String.join(" ", encodedImgPixelAlpha)).getBytes());
                outputStream.write(separator.getBytes());
                outputStream.write(("Encoded Image Pixel (Full): " + encodedImgRgbBinStr).getBytes());
                outputStream.write(separator.getBytes());

                final String retrievedBin = encodedImgPixelRed.get(1) + encodedImgPixelGreen.get(1) + encodedImgPixelBlue.get(1);

                if (!bufferRetrieved) {
                    if (retrievedBin.equals(String.join("", delimiterList))) {
                        bufferRetrieved = true;
                        retrievedImgHeight = new BigInteger(retrievedBuffer.toString(), 2).intValue();

                        outputStream.write(("Width: " + retrievedImgWidth + ", Height: " + retrievedImgHeight + separator + separator).getBytes());
                        retrievedBufferedImg = new BufferedImage(retrievedImgWidth, retrievedImgHeight, encodedBufferedImg.getType());
                    }
                    else if (retrievedBin.equals(String.join("", bufferDelimiterList))) {
                        retrievedImgWidth = new BigInteger(retrievedBuffer.toString(), 2).intValue();
                        retrievedBuffer = new StringBuilder();
                    }
                    else
                        retrievedBuffer.append(retrievedBin);
                }
                else {
                    if (retrievedBin.equals(String.join("", delimiterList)))
                        break outer;
                    else {
                        final String retrievedImgPixelRed = String.join("", new ArrayList<String>() {{ add(encodedImgPixelRed.get(1)); add(padding); }});
                        final String retrievedImgPixelGreen = String.join("", new ArrayList<String>() {{ add(encodedImgPixelGreen.get(1)); add(padding); }});
                        final String retrievedImgPixelBlue = String.join("", new ArrayList<String>() {{ add(encodedImgPixelBlue.get(1)); add(padding); }});
                        final String retrievedImgPixelAlpha = String.join("", new ArrayList<String>() {{ add(encodedImgPixelAlpha.get(1)); add(padding); }});
                        final String retrievedImgRgbBinStr = retrievedImgPixelAlpha + retrievedImgPixelRed + retrievedImgPixelGreen + retrievedImgPixelBlue;

                        outputStream.write(("Retrieved Image Pixel (x:" + retrievedImgXIndex + ", y:" + retrievedImgYIndex + "): R - " + String.join(" ", StringManipulations.chunkString(retrievedImgPixelRed, 4)) + " G - " + String.join(" ", StringManipulations.chunkString(retrievedImgPixelGreen, 4)) + " B - " + String.join(" ", StringManipulations.chunkString(retrievedImgPixelBlue, 4)) + " A - " + String.join(" ", StringManipulations.chunkString(retrievedImgPixelAlpha, 4))).getBytes());
                        outputStream.write(separator.getBytes());
                        outputStream.write(("Retrieved Image Pixel (Full): " + String.join(" | ", StringManipulations.chunkString(retrievedImgRgbBinStr, 8))).getBytes());

                        outputStream.write(separator.getBytes());
                        outputStream.write(separator.getBytes());

                        retrievedBufferedImg.setRGB(retrievedImgXIndex, retrievedImgYIndex, (new BigInteger(retrievedImgRgbBinStr, 2).intValue()));
                    }

                    if (retrievedImgXIndex + 1 == retrievedImgWidth) {
                        retrievedImgXIndex = 0;
                        retrievedImgYIndex++;
                    }
                    else
                        retrievedImgXIndex++;
                }
            }
        }

        outputStream.flush();
        outputStream.close();

        return retrievedBufferedImg;
    }

    public static class OriginalPicTooSmallException extends Exception {
        public String getMessage() {
            return "Original Picture is too small!";
        }

        public String toString() {
            return this.getClass().getCanonicalName() + ": " + this.getMessage();
        }
    }
}

File: kotlin/steganography(pic_in_pic)/src/SteganographyPic.kt
/*
* Created By: Jose A. Alvarado
* */

import java.awt.Color
import java.awt.image.BufferedImage
import java.io.File
import java.io.FileOutputStream
import java.lang.StringBuilder
import java.math.BigInteger
import java.util.*
import javax.imageio.ImageIO
import kotlin.math.ceil

class SteganographyPic {
    companion object {
        private val bufferDelimiterList = listOf("1010", "1111", "0101")
        private val delimiterList = listOf("1111", "1010", "1111")
        private const val padding = "0000"
        private const val bufferGrouping = 3
        private val separator = System.lineSeparator()

        fun generateImage(image : BufferedImage, picturePath : String) : String  {
            val picturePathFile = File(picturePath)

            ImageIO.write(image, "png", picturePathFile)
            return picturePathFile.absolutePath
        }

        fun completeGroupsForPixelEmbedding(input: Int) : List<String> {
            val binInput = Integer.toBinaryString(input)
            val paddingAmt = ceil(binInput.length / 4.0).toInt() * 4

            val currGroupAmt = paddingAmt / 4
            val targetGroupAmt = ceil(currGroupAmt / (bufferGrouping + 0.0)).toInt() * bufferGrouping
            val neededGroupAmt = targetGroupAmt - currGroupAmt

            var overallGroupPadding = paddingAmt
            if (overallGroupPadding % bufferGrouping != 0)
                overallGroupPadding =
                    ceil(overallGroupPadding / (bufferGrouping + 0.0) + neededGroupAmt).toInt() * bufferGrouping

            return binInput.padStart(overallGroupPadding, '0').chunked(4)
        }

        @Throws(OriginalPicTooSmallException::class)
        fun embedMessage(pictureToEncode : BufferedImage, pictureToHide: BufferedImage) : BufferedImage {
            val startTime = Date()

            val originalBufferedImg: BufferedImage = pictureToEncode
            val originalPicWidth: Int = originalBufferedImg.width
            val originalPicHeight: Int = originalBufferedImg.height
            val originalPicPixelArea: Int = originalPicHeight * originalPicWidth

            val picToHideBufferedImage: BufferedImage = pictureToHide
            val picToHideWidth: Int = picToHideBufferedImage.width
            val picToHideHeight: Int = picToHideBufferedImage.height
            val picToHidePixelArea: Int = picToHideWidth * picToHideHeight

            val picToHideWidthBinList = completeGroupsForPixelEmbedding(picToHideWidth)
            val picToHideHeightList = completeGroupsForPixelEmbedding(picToHideHeight)
            val picToHidePixelAreaBinList = picToHideWidthBinList + bufferDelimiterList + picToHideHeightList
            val totalPixelsForBuffer = picToHidePixelAreaBinList.size / bufferGrouping

            if (picToHidePixelArea > (originalPicPixelArea - totalPixelsForBuffer))
                throw OriginalPicTooSmallException()

            val newEncodedBufferedImage = BufferedImage(originalBufferedImg.colorModel, originalBufferedImg.raster, originalBufferedImg.isAlphaPremultiplied, null)

            val file = File(".\\debugEn.txt")
            val outputStream = FileOutputStream(file)

            outputStream.write("Image in Binary...$separator$separator".toByteArray())

            var pixelsLeftForBuffer = totalPixelsForBuffer
            var insertDelimiter = false
            var picToHideYIndex = 0
            var picToHideXIndex = -2
            var shouldEmbedPixel = true

            outer@ for (yIndex in (0 until originalPicHeight)) {
                for (xIndex in (0 until originalPicWidth)) {
                    val originalImgPixel = originalBufferedImg.getRGB(xIndex, yIndex)

                    if (shouldEmbedPixel) {

                        val originalImgPixelColor = Color(originalImgPixel)
                        val originalImgPixelRed = Integer.toBinaryString(originalImgPixelColor.red).padStart(8, '0').chunked(4)
                        val originalImgPixelGreen = Integer.toBinaryString(originalImgPixelColor.green).padStart(8, '0').chunked(4)
                        val originalImgPixelBlue = Integer.toBinaryString(originalImgPixelColor.blue).padStart(8, '0').chunked(4)
                        val originalImgPixelAlpha = Integer.toBinaryString(originalImgPixelColor.alpha).padStart(8, '0').chunked(4)
                        val originalImgRgbBinStr = Integer.toBinaryString(originalImgPixel).chunked(8).joinToString(" | ") { it.chunked(4).joinToString(" ") }

                        outputStream.write("Original Image Pixel (x:$xIndex, y:$yIndex): R - ${originalImgPixelRed.joinToString(" ")}, G - ${originalImgPixelGreen.joinToString(" ")}, B - ${originalImgPixelBlue.joinToString(" ")}, A - ${originalImgPixelAlpha.joinToString(" ")}".toByteArray())
                        outputStream.write(separator.toByteArray())
                        outputStream.write("Original Image Pixel (Full): $originalImgRgbBinStr".toByteArray())

                        outputStream.write(separator.toByteArray())

                        val picToHidePixelRed: List<String>
                        val picToHidePixelGreen: List<String>
                        val picToHidePixelBlue: List<String>
                        val picToHidePixelAlpha: List<String>
                        val picToHideRgbBinStr: String

                        if (pixelsLeftForBuffer > 0) {
                            val startingIndex = picToHidePixelAreaBinList.size - (pixelsLeftForBuffer * bufferGrouping)

                            picToHidePixelRed = listOf(picToHidePixelAreaBinList[startingIndex], padding)
                            picToHidePixelGreen = listOf(picToHidePixelAreaBinList[startingIndex + 1], padding)
                            picToHidePixelBlue = listOf(picToHidePixelAreaBinList[startingIndex + 2], padding)
                            picToHidePixelAlpha = originalImgPixelAlpha
                            picToHideRgbBinStr = picToHidePixelAlpha.joinToString(" ") + " | " + picToHidePixelRed.joinToString(" ") + " | " + picToHidePixelGreen.joinToString(" ") + " | " + picToHidePixelBlue.joinToString(" ")

                            outputStream.write("BUFFER $separator".toByteArray())

                            pixelsLeftForBuffer--

                            if (pixelsLeftForBuffer == 0)
                                insertDelimiter = true
                        }
                        else if (insertDelimiter) {
                            picToHidePixelRed = listOf(delimiterList[0], delimiterList[0])
                            picToHidePixelGreen = listOf(delimiterList[1], delimiterList[1])
                            picToHidePixelBlue = listOf(delimiterList[2], delimiterList[2])
                            picToHidePixelAlpha = originalImgPixelAlpha
                            picToHideRgbBinStr = picToHidePixelAlpha.joinToString(" ") + " | " +  picToHidePixelRed.joinToString(" ") + " | " + picToHidePixelGreen.joinToString(" ") + " | " + picToHidePixelBlue.joinToString(" ")

                            outputStream.write("DELIMITER $separator".toByteArray())
                            insertDelimiter = false
                        }
                        else {
                            val picToHidePixel = picToHideBufferedImage.getRGB(picToHideXIndex, picToHideYIndex)
                            val picToHidePixelColor = Color(picToHidePixel)

                            picToHidePixelRed = Integer.toBinaryString(picToHidePixelColor.red).padStart(8, '0').chunked(4)
                            picToHidePixelGreen = Integer.toBinaryString(picToHidePixelColor.green).padStart(8, '0').chunked(4)
                            picToHidePixelBlue = Integer.toBinaryString(picToHidePixelColor.blue).padStart(8, '0').chunked(4)
                            picToHidePixelAlpha = Integer.toBinaryString(picToHidePixelColor.alpha).padStart(8, '0').chunked(4)
                            picToHideRgbBinStr = Integer.toBinaryString(picToHidePixel).chunked(8).joinToString(" | ") { it.chunked(4).joinToString(" ") }
                        }

                        outputStream.write("Picture To Hide Image Pixel (x:$picToHideXIndex, y:$picToHideYIndex): R - ${picToHidePixelRed.joinToString(" ")}, G - ${picToHidePixelGreen.joinToString(" ")}, B - ${picToHidePixelBlue.joinToString(" ")}, A - ${picToHidePixelAlpha.joinToString(" ")}".toByteArray())
                        outputStream.write(separator.toByteArray())
                        outputStream.write("Picture To Hide Image Pixel (Full): $picToHideRgbBinStr".toByteArray())

                        outputStream.write(separator.toByteArray())

                        val newImgPixelRed = listOf(originalImgPixelRed[0], picToHidePixelRed[0]).joinToString("")
                        val newImgPixelGreen = listOf(originalImgPixelGreen[0], picToHidePixelGreen[0]).joinToString("")
                        val newImgPixelBlue = listOf(originalImgPixelBlue[0], picToHidePixelBlue[0]).joinToString("")
                        val newImgPixelAlpha = listOf(originalImgPixelAlpha[0], picToHidePixelAlpha[0]).joinToString("")
                        val newImgRgbBinStr = newImgPixelAlpha + newImgPixelRed + newImgPixelGreen + newImgPixelBlue

                        outputStream.write("New Image Pixel (x:$xIndex, y:$yIndex): R - ${newImgPixelRed.chunked(4).joinToString(" ")}, G - ${newImgPixelGreen.chunked(4).joinToString(" ")}, B - ${newImgPixelBlue.chunked(4).joinToString(" ")}, A - ${newImgPixelAlpha.chunked(4).joinToString(" ")}".toByteArray())
                        outputStream.write(separator.toByteArray())
                        outputStream.write("New Image Pixel (Full): ${newImgRgbBinStr.chunked(8).joinToString(" | ") { it.chunked(4).joinToString(" ") }}".toByteArray())


                        outputStream.write(separator.toByteArray())
                        outputStream.write(separator.toByteArray())

                        newEncodedBufferedImage.setRGB(xIndex, yIndex, BigInteger(newImgRgbBinStr, 2).toInt())
                    }
                    else {
                        //newEncodedBufferedImage.setRGB(xIndex, yIndex, originalImgPixel)
                        break@outer
                    }

                    if (picToHideYIndex + 1 == picToHideHeight && picToHideXIndex + 1 == picToHideWidth) {
                        insertDelimiter = true
                        picToHideYIndex = -100
                    }
                    else if (picToHideYIndex < 0)
                        shouldEmbedPixel = false

                    if (picToHideXIndex + 1 == picToHideWidth) {
                        picToHideXIndex = 0
                        picToHideYIndex++
                    }
                    else if (pixelsLeftForBuffer == 0)
                        picToHideXIndex++
                }
            }

            outputStream.flush()
            outputStream.close()

            val endTime = Date()
            val timeDiff = (endTime.time - startTime.time) + 0.0
            println("\nEncoding Time: $timeDiff milliseconds\n\n")

            return newEncodedBufferedImage
        }

        fun retrieveEncodedImageFromImage(encodedPic: BufferedImage) : BufferedImage {
            val startTime = Date()

            var retrievedImgHeight = 100
            var retrievedImgWidth = 100

            val encodedBufferedImg = encodedPic
            var retrievedBufferedImg = BufferedImage(retrievedImgWidth, retrievedImgHeight, encodedBufferedImg.type)

            var bufferRetrieved = false
            val retrievedBuffer = StringBuilder("")
            var retrievedImgYIndex = 0
            var retrievedImgXIndex = 0

            val file = File(".\\debugRet.txt")
            val outputStream = FileOutputStream(file)

            outer@ for (yIndex in (0 until encodedBufferedImg.height)) {
                for (xIndex in (0 until encodedBufferedImg.width)) {
                    val encodedImgPixel = encodedBufferedImg.getRGB(xIndex, yIndex)
                    val encodedImgPixelColor = Color(encodedImgPixel)
                    val encodedImgPixelRed = Integer.toBinaryString(encodedImgPixelColor.red).padStart(8, '0').chunked(4)
                    val encodedImgPixelGreen = Integer.toBinaryString(encodedImgPixelColor.green).padStart(8, '0').chunked(4)
                    val encodedImgPixelBlue = Integer.toBinaryString(encodedImgPixelColor.blue).padStart(8, '0').chunked(4)
                    val encodedImgPixelAlpha = Integer.toBinaryString(encodedImgPixelColor.alpha).padStart(8, '0').chunked(4)
                    val encodedImgRgbBinStr = Integer.toBinaryString(encodedImgPixel).chunked(8).joinToString(" | ") { it.chunked(4).joinToString(" ") }

                    outputStream.write("Encoded Image Pixel (x:$xIndex, y:$yIndex): R - ${encodedImgPixelRed.joinToString(" ")}, G - ${encodedImgPixelGreen.joinToString(" ")}, B - ${encodedImgPixelBlue.joinToString(" ")}, A - ${encodedImgPixelAlpha.joinToString(" ")}".toByteArray())
                    outputStream.write(separator.toByteArray())
                    outputStream.write("Encoded Image Pixel (Full): $encodedImgRgbBinStr".toByteArray())
                    outputStream.write(separator.toByteArray())

                    val retrievedBin = encodedImgPixelRed[1] + encodedImgPixelGreen[1] + encodedImgPixelBlue[1]

                    if (!bufferRetrieved) {
                        when (retrievedBin) {
                            delimiterList.joinToString("") -> {
                                bufferRetrieved = true
                                retrievedImgHeight = BigInteger(retrievedBuffer.toString(), 2).toInt()

                                outputStream.write("Width: $retrievedImgWidth, Height: $retrievedImgHeight $separator$separator".toByteArray())
                                retrievedBufferedImg = BufferedImage(retrievedImgWidth, retrievedImgHeight, encodedBufferedImg.type)
                            }
                            bufferDelimiterList.joinToString("") -> {
                                retrievedImgWidth = BigInteger(retrievedBuffer.toString(), 2).toInt()
                                retrievedBuffer.clear()
                            }
                            else -> retrievedBuffer.append(retrievedBin)
                        }
                    }
                    else {
                        if (retrievedBin == delimiterList.joinToString("")) {
                            break@outer
                        }
                        else {
                            val retrievedImgPixelRed = listOf(encodedImgPixelRed[1], padding).joinToString("")
                            val retrievedImgPixelGreen = listOf(encodedImgPixelGreen[1], padding).joinToString("")
                            val retrievedImgPixelBlue = listOf(encodedImgPixelBlue[1], padding).joinToString("")
                            val retrievedImgPixelAlpha = listOf(encodedImgPixelAlpha[1], "1111").joinToString("")
                            val retrievedImgRgbBinStr = retrievedImgPixelAlpha + retrievedImgPixelRed + retrievedImgPixelGreen + retrievedImgPixelBlue

                            outputStream.write("Retrieved Image Pixel (x:$retrievedImgXIndex, y:$retrievedImgYIndex): R - ${retrievedImgPixelRed.chunked(4).joinToString(" ")}, G - ${retrievedImgPixelGreen.chunked(4).joinToString(" ")}, B - ${retrievedImgPixelBlue.chunked(4).joinToString(" ")}, A - ${retrievedImgPixelAlpha.chunked(4).joinToString(" ")}".toByteArray())
                            outputStream.write(separator.toByteArray())
                            outputStream.write("Retrieved Image Pixel (Full): ${retrievedImgRgbBinStr.chunked(8).joinToString(" | ") { it.chunked(4).joinToString(" ") }}".toByteArray())


                            outputStream.write(separator.toByteArray())
                            outputStream.write(separator.toByteArray())

                            retrievedBufferedImg.setRGB(retrievedImgXIndex, retrievedImgYIndex, BigInteger(retrievedImgRgbBinStr, 2).toInt())
                        }

                        if (retrievedImgXIndex + 1 == retrievedImgWidth) {
                            retrievedImgXIndex = 0
                            retrievedImgYIndex++
                        }
                        else
                            retrievedImgXIndex++
                    }
                }
            }

            outputStream.flush()
            outputStream.close()

            val endTime = Date()
            val timeDiff = (endTime.time - startTime.time) + 0.0
            println("\nDecoding Time: $timeDiff milliseconds\n\n")

            return retrievedBufferedImg
        }

        class OriginalPicTooSmallException : Exception() {
            val errMsg = "Original Picture is too small!"
        }
    }
}